main.cpp

```diff
@@ -0,0 +1,193 @@
+  glfwSetKeyCallback(mWindow, key_callback);
+  
+  // Create Context and Load OpenGL Functions
+  glfwMakeContextCurrent(mWindow);
+  gladLoadGL();
+  fprintf(stderr, "OpenGL %s\n", glGetString(GL_VERSION));
+  
+  // Create Vertex Array Object: this will store all the information about the vertex data that we are about to specify
+  GLuint vao;
+  glGenVertexArrays(1, &vao);
+  glBindVertexArray(vao);
+  
+  // Create a Vertex Buffer Object and copy the vertex data to it
+  GLuint vbo;
+  glGenBuffers(1, &vbo);
+  glBindBuffer(GL_ARRAY_BUFFER, vbo);
+  glBufferData(GL_ARRAY_BUFFER, sizeof(vertices), vertices, GL_STATIC_DRAW);
+  
+  // INDEX BUFFER
+  GLuint IBO;
+  unsigned int Indices[] = { 0, 1, 2,
+                             1, 2, 3,
+                             2, 3, 0,
+                             3, 0, 1 };
+						   
+  glGenBuffers(1, &IBO);
+  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, IBO);
+  glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(Indices), Indices, GL_STATIC_DRAW);
+
+  // Create and compile the vertex shader
+  GLuint vertexShader = glCreateShader(GL_VERTEX_SHADER);
+  glShaderSource(vertexShader, 1, &vertexSource, NULL);
+  glCompileShader(vertexShader);
+  // Create and compile the fragment shader
+  GLuint fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
+  glShaderSource(fragmentShader, 1, &fragmentSource, NULL);
+  glCompileShader(fragmentShader);
+  // Link the vertex and fragment shader into a shader program
+  GLuint shaderProgram = glCreateProgram();
+  glAttachShader(shaderProgram, vertexShader);
+  glAttachShader(shaderProgram, fragmentShader);
+  glBindFragDataLocation(shaderProgram, 0, "outColor");
+  glLinkProgram(shaderProgram);
+  glUseProgram(shaderProgram);
+  
+  // Specify the layout of the vertex data
+  // position
+  GLint posAttrib = glGetAttribLocation(shaderProgram, "position");
+  glEnableVertexAttribArray(posAttrib);
+  glVertexAttribPointer(posAttrib, 3, GL_FLOAT, GL_FALSE, 6 * sizeof(GLfloat), 0);  // attribute location, # values, value type, normalize?, stride, offset
+  // color
+  GLint colAttrib = glGetAttribLocation(shaderProgram, "color");
+  glEnableVertexAttribArray(colAttrib);
+  glVertexAttribPointer(colAttrib, 3, GL_FLOAT, GL_FALSE, 6 * sizeof(GLfloat), (void*)(3 * sizeof(GLfloat)));
+  
+  // model matrix
+  GLint modelTransform = glGetUniformLocation(shaderProgram, "model");
+  glm::mat4 scale_model = glm::scale(glm::mat4(1.0f), glm::vec3(0.5f,0.5f,0.5f));
+	glm::mat4 model = scale_model;
+  glUniformMatrix4fv(modelTransform, 1, GL_FALSE, glm::value_ptr(model));
+  
+  // Rendering Loop
+  
+  glDepthFunc(GL_LESS);
+  glEnable(GL_DEPTH_TEST);
+  
+  while (glfwWindowShouldClose(mWindow) == false) {
+    
+    // Background Fill Color
+    glClearColor(0.85f, 0.65f, 0.65f, 0.8f);
+    glClearDepth(1.0f);
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+    
+    // draw triangle
+    glDrawElements(GL_TRIANGLES, 12, GL_UNSIGNED_INT, 0);
+	
+	for(int i = 0; i < 4; i++) {

```

Comment:
this is inefficient. You should use glm or create a single transformation matrix and send it once to the shader so it can be multiplied by all vertices in the GPU instead of looping over all vertices in the application.